More Linux fixes

robxnano opened the pull request:
**Description of Change:**
I've had some more time to fix a few of the outstanding bugs in the Linux UI. I also added some new features:
- Ability to drag and drop, and select multiple items in the queue
- Full screen toggle button in preview window

Hopefully this will help make the interface more consistent with the other versions. As usual, if there are any problems, just ask and I'll be happy to make the necessary changes.

**Test on:**

- [x] Ubuntu 22.04
- [x] Fedora 37 KDE
- [x] Linux Mint 21

robxnano committed:
LinGUI: Add setting to disable preview on Summary tab

Fixes #1267.

robxnano committed:
LinGUI: Fixes for preview window

- Fix for preview being bigger than the screen on HIDPI displays
(Fixes https://github.com/HandBrake/HandBrake/issues/4126)
- Also fixes similar issues when encoding a video with a
higher resolution than the screen on normal DPI displays
- Added fullscreen mode toggle button
- Can be activated by double clicking on the window
- Improved styling of overlay controls
- Removed unneeded code

robxnano committed:
LinGUI: Clean up UI files and fix inconsistencies

- Updated syntax in ghb3.ui
- Added titles to dialogs that didn't have one
- Added default actions to dialogs so that Enter key works

robxnano committed:
LinGUI: Add Move To Top/Bottom to queue menu

As the Linux GUI doesn't have a queue context menu yet, this can
be accessed from the Options menu at the top. Closes #3065.

robxnano committed:
LinGUI: Add queue context menu, allow multi-select

- Multiple items can be selected at once by holding Ctrl/Shift
- Drag and drop works even when multiple items selected
- Context menu items apply to all selected items, if applicable
- Added Play File option to context menu

robxnano committed:
LinGUI: Add subtitle and audio track names to queue #4155

robxnano committed:
LinGUI: Restore ability to scan folders

robxnano committed:
LinGUI: Add Audio Gain to queue summary

Fixes #4714

robxnano committed:
Flatpak: Replace x11 permission with fallback-x11

The fallback permission allows X11 to be used when Wayland isn't
available. As a GTK3 appplication, HandBrake is already running
under Wayland whenever possible, so this will not make any practical
difference apart from removing an unnecessary permission warning.

sr55 commented:
Sweet. Nice collection of changes. I'll run some tests over the weekend and get feedback to you.

Thanks for taking the time to give the Linux UI some love!

sr55 commented:
Closing in favour of #4720 since it has all these commits as well.

galad87 commented:
I'm reopening this, in the case the native file picker needs additional testing.
I did a quick test and it seems to work ok, apart from the "Source resolution" button in the preview window: the button is visible the first time you open the window, and clicking it just make it disappear because the window it's already the right size, and if I maximise he window click the button does not nothing, and it just disappear again.

ferdnyc commented on gtk/src/ghb3.ui:
```diff
@@ -8989,7 +8994,7 @@ Uncheck this if you want to allow changing each title's settings independently.<
                             <property name="visible">True</property>
                             <property name="can_focus">False</property>
-                            <property name="icon_name">gtk-media-play</property>
+                            <property name="icon_name">media-playbck-start</property>
```
I don't know why GitHub is being screwy on me — I can't view this file diff in the full set of PR changes, and I can't leave a suggested change with this comment — but:

Typo. this should be:

```xml
                            <property name="icon_name">media-playback-start</property>
```
(Or, after the changes in the next commit...)
```xml
                            <property name="icon-name">media-playback-start</property>
```

ferdnyc commented on gtk/src/ghb3.ui:
```diff
@@ -9306,7 +9306,7 @@ Uncheck this if you want to allow changing each title's settings independently.<
     <action-widgets>
       <action-widget response="cancel">source_cancel</action-widget>
-      <action-widget response="ok">source_ok</action-widget>
+      <action-widget response="no">source_ok</action-widget>
```
Should this be `response="yes"`, instead?  an Open button with the ID `source_ok`  that sends a `no` response is hurting my brain. (Also, maybe the ID should be `source_open`?)

ferdnyc replied:
Or, you could use `response="accept"` or `response="reject"`, which according to the docs are both a:
> Generic response id, not used by GTK+ dialogs

But either way I'd definitely change the widget `id`.

robxnano replied:
This is what I thought originally, which is why I changed it [previously](https://github.com/HandBrake/HandBrake/commit/802caf8a78dfd31ee2695bb1b397067bcc17277b). But then I found that it was done deliberately, in order to let you select a folder to scan instead of just a single file.

> For most dialog boxes you can use your own custom response codes rather than the ones in GtkResponseType, but GtkFileChooserDialog assumes that its “accept”-type action, e.g. an “Open” or “Save” button, will have one of the following response codes:
> 
> GTK_RESPONSE_ACCEPT
> 
> GTK_RESPONSE_OK
> 
> GTK_RESPONSE_YES
> 
> GTK_RESPONSE_APPLY
> 
> This is because GtkFileChooserDialog must intercept responses and switch to folders if appropriate, rather than letting the dialog terminate — the implementation uses these known response codes to know which responses can be blocked if appropriate.

Basically if you select a folder and press Open, the dialog closes, thinking it was cancelled, but HandBrake still gets the selected folder and tries to open it. It may not be very intuitive, but it saves having a separate menu item for opening a folder. Having said that, using a non-standard code (but not `accept`) also seems to work, so I could change it to that instead. 

Also, in my other pull request which switches the file dialogs to GtkFileChooserNative, I had to add an Open Directory menu item anyway as it doesn't let you set the response codes. So this could end up being irrelevant soon.

ferdnyc replied:
Ah! That makes sense. (Insofar as anything in Gtk makes sense, I mean. It's all relative.)

No harm in using "no" then, IMHO, though It'd definitely be less brain-straining if the widget ID was changed from `source_ok` to `source_open`. (Maybe even with a `<!-- … -->` comment in the `.ui` file, explaining that `source_open` needs to send a `no` signal to fake folder-selection semantics.)

ferdnyc commented:
@galad87 
> I did a quick test and it seems to work ok, apart from the "Source resolution" button in the preview window: the button is visible the first time you open the window, and clicking it just make it disappear because the window it's already the right size, and if I maximise he window click the button does not nothing, and it just disappear again.

It's worth noting that the preview window sizing is _already_ kind of screwy in Gtk, at least under Wayland. (Mostly due to the attempts to constrain aspect ratio, which are only partially effective.)

With fullscreening/maximizing, there's a question of what pressing "Source Resolution" _should_ do:
1. Implicitly un-fullscreen/un-maximize the window? (That was the previous behavior with maximization, though fullscreening feels like a potentially different situation.)
2. Reduce the size of the preview **widget**, without affecting the window size? 
3. Nothing, if fullscreened? (That would be more in line with e.g. fullscreened VLC, where the Video > Zoom menu no longer has any effect.) In which case, maybe the "Source Resolution" button should be disabled in fullscreen.
4. Some other hybrid behavior, e.g. (1) if maximized, (2) if fullscreened?

ferdnyc commented on gtk/src/queuehandler.c:
```diff
@@ -2578,15 +2604,58 @@ queue_key_press_cb(
+        event_win = gdk_window_get_effective_parent (event_win);
+    }
+
+    if (event->type == GDK_2BUTTON_PRESS)
```
`GDK_2BUTTON_PRESS` and `GDK_3BUTTON_PRESS` were removed from Gdk 4.0, so this will break Gtk4 compatibility.

ferdnyc replied:
I mean, not that it's the _only_ thing. All access to event internals [is disabled in Gtk4](https://docs.gtk.org/gtk4/migrating-3to4.html#adapt-to-gdkevent-api-changes):

> ### Adapt to `GdkEvent` API changes
> Direct access to [GdkEvent](https://docs.gtk.org/gdk4/class.Event.html) structs is no longer possible in GTK 4. GdkEvent is now a strictly read-only type, and you can no longer change any of its fields, or construct new events. All event fields have accessors that you will have to use.
> 
> Event compression is always enabled in GTK 4, for both motion and scroll events. If you need to see the uncoalesced motion or scroll history, use [gdk_event_get_history()](https://docs.gtk.org/gdk4/method.Event.get_history.html) on the latest event.

So as I'm sure you're aware, a _lot_ of the `master` branch code in `gtk/src/callbacks.c`, `gtk/src/queuehandler.c`, and `gtk/src/ghbcompat.h` breaks under Gtk4. But adding new `::button-press-event` handlers and uses of <code>GTK_<var>(N)</var>BUTTON_PRESS</code> will only further move the codebase away from being able to support Gtk4. For that, it needs to:

> ### [Stop using `GtkWidget` event signals](https://docs.gtk.org/gtk4/migrating-3to4.html#stop-using-gtkwidget-event-signals)
> Event controllers and gestures replace event signals in GTK 4.
> 
> Most of them have been backported to GTK 3.x so you can prepare for this change.
> 
> |Signal	|Event controller|
> |-----------|-----------------------|
> |::event	|GtkEventControllerLegacy|
> |::event-after	|GtkEventControllerLegacy|
> |::button-press-event	|GtkGestureClick|
> |::button-release-event	|GtkGestureClick
> |::touch-event	|various touch gestures
> |::scroll-event	|GtkEventControllerScroll
> |::motion-notify-event	|GtkEventControllerMotion
> |::delete-event	|-
> |::key-press-event	|GtkEventControllerKey
> |::key-release-event	|GtkEventControllerKey
> |::enter-notify-event	|GtkEventControllerMotion
> |::leave-notify-event	|GtkEventControllerMotion
> |::configure-event	|-
> |::focus-in-event	|GtkEventControllerFocus
> |::focus-out-event	|GtkEventControllerFocus
> |::map-event	|-
> |::unmap-event	|-
> |::property-notify-event	|replaced by GdkClipboard
> |::selection-clear-event	|replaced by GdkClipboard
> |::selection-request-event	|replaced by GdkClipboard
> |::selection-notify-event	|replaced by GdkClipboard
> |Drag-and-Drop signals	|GtkDragSource, GtkDropTarget
> |::proximity-in-event	|GtkGestureStylus
> |::proximity-out-event	|GtkGestureStylus
> |::visibility-notify-event	|-
> |::window-state-event	|-
> |::damage-event	|-
> |::grab-broken-event	|-
> 
> Event signals which are not directly related to input have to be dealt with on a one-by-one basis:
> 
> * If you were using `::configure-event` and `::window-state-event` to save window state, you should use property notification for corresponding `GtkWindow` properties, such as `GtkWindow:default-width`, `GtkWindow:default-height`, `GtkWindow:maximized` or `GtkWindow:fullscreened`.
> * If you were using `::delete-event` to present a confirmation when using the close button of a window, you should use the `GtkWindow::close-request` signal.
> * If you were using `::map-event` and `::unmap-event` to track a window being mapped, you should use property notification for the `GdkSurface:mapped` property instead.
> * The `::damage-event` signal has no replacement, as the only consumer of damage events were the offscreen GDK surfaces, which have no replacement in GTK 4.x.
>

ferdnyc replied:
(For multi-clicks, specifically, [`GtkGestureClick::pressed`](https://docs.gtk.org/gtk4/signal.GestureClick.pressed.html) includes an `n_press` parameter that lists how many times the button was clicked.)

ferdnyc replied:
Of course, despite their reassurances about backporting, it seems `GtkGestureClick` _hasn't_ been backported to Gtk3, so in reality there will most likely still need to be some version-based alternation that either uses an event callback and `GTK_{2,3}BUTTON_PRESS` or a `GtkGesture` event controller that accesses the underlying event on Gtk3, and `GtkGestureClick` on Gtk4. Whether that happens right in the handlers, or in `ghbcompat.h`... sigh.

So, I guess, "Neeeeevermind." `¯\_(ツ)_/¯`

ferdnyc replied:
Oh, now I see! It's called [`GtkGestureMultiPress`](https://docs.gtk.org/gtk3/class.GestureMultiPress.html) in Gtk3. (I thought that was strictly a multi-touch gesture controller, but nope, it apparently recognizes multiple button clicks too, and has the same `n_press` property in its `::pressed` signal.)

robxnano replied:
Thanks for this. I must have assumed that the new APIs were only available in GTK4, but it seems they've been available for a while. I've replaced all the button press events with a function in ghbcompat.h that uses GtkGestureMultiPress or GtkGestureClick depending on the version. I don't know if the GTK4 code actually works, due to all the other problems, but if not it should be easy enough to fix once everything else works.

ferdnyc commented:
@robxnano 

Hmm... one issue with the folder-selection thing — because the `Open` button in the file selection dialog is has `receives-default` and `has-default`, things like pressing <kbd>Enter</kbd> or <kbd>Space</kbd> with a folder highlighted in the tree will select that folder, instead of descending into it. Basically, it breaks keyboard navigation of the folder tree.

robxnano committed:
LinGUI: Use GtkGestureMultiPress

Replace button-press-event for multi-click and right-click actions
to ensure compatibility with GTK4.

robxnano committed:
LinGUI: Change response id for source dialog

- Clarify meaning by using custom response id of 1 instead of
GTK_RESPONSE_NO for Open action
- Fix typo

robxnano commented:
> It's worth noting that the preview window sizing is _already_ kind of screwy in Gtk, at least under Wayland. (Mostly due to the attempts to constrain aspect ratio, which are only partially effective.)
> 
> With fullscreening/maximizing, there's a question of what pressing "Source Resolution" _should_ do:
> 
>     1. Implicitly un-fullscreen/un-maximize the window? (That was the previous behavior with maximization, though fullscreening feels like a potentially different situation.)
> 
>     2. Reduce the size of the preview **widget**, without affecting the window size?
> 
>     3. Nothing, if fullscreened? (That would be more in line with e.g. fullscreened VLC, where the Video > Zoom menu no longer has any effect.) In which case, maybe the "Source Resolution" button should be disabled in fullscreen.
> 
>     4. Some other hybrid behavior, e.g. (1) if maximized, (2) if fullscreened?

There's also the 'Scale down High Definition previews' option to take into account. If that's enabled, the preview is limited to 90% of the height/width of the screen (unless maximized/full screen) to avoid having a window which doesn't fit on the screen. Maybe if that's the case the 'Source Resolution' button should always be hidden, since it can never really do what it claims to? But then users might be confused that the button doesn't appear on certain videos. Yeah, it's complicated.

robxnano committed:
LinGUI: Removed page transition animations

robxnano committed:
LinGUI: Fix resizing preview window

- Remove buggy aspect ratio constraint on resizing
- Allow window to be resized regardless of the original video size
- Fix most cases of the Source Resolution button showing up wrongly
- Allow the Source Resolution button to unmaximize the window

robxnano committed:
Fix modelines to match the actual indentation used

galad87 commented:
The Mac UI has button "Scale to screen - Real size" toggle button, and if the window is in full screen mode it just rescale the preview image inside the window.
Which might not be perfect, it's just the way it works now.

robxnano committed:
LinGUI: Don't ask to restart when not needed

If the custom temp dir option was enabled, HandBrake would ask to
restart every time the preferences dialog was closed, even if the
option hadn't been changed.

ferdnyc commented:
@robxnano

Just built the latest changes, the gesture handlers work well! Thanks for doing that, (restored) Gtk4 compatibility at least feels more _within reach_ now.

Completely tangential: I wonder if it would be worth it to break `gtk/src/ghb3.ui` into separate files for the different windows? (Main, Preferences, Preview, Queue... did I miss any?) Mostly because it's become sort of unnavigable, particularly in GitHub's PR interface which refuses to actually render diffs of the changes for me. It is kind of an unwieldy beast, having everything crammed in together like that. (The file is over 10k lines, after all!)

One _huge_ improvement I notice already, with these changes vs. `master`: although preview window sizing is still screwy and a bit broken (I'll detail exactly how in another comment), on your branch when the window mis-sizes itself, _at least_ there's a black background filling the spaces where the window area exceeds the preview dimensions. In the release/`master` code, frequently _nothing_ is drawn into those empty spaces, leaving the preview displayed over a sort of twitchily-transparent Nothing that's really hard on the eyes.

ferdnyc commented:
@robxnano 

So, here's how Preview plays out for me (for the most part, consistently/reproducibly), on Fedora 36 under GNOME Shell on Wayland, with a 720p HD video loaded into HandBrake running on my 24" 1920×1200 (16:10) desktop display.

1. Open Preview.
2. Initially, I can resize the window larger than the native resolution but not smaller. (Aside: ...why not?)
    1. Fortunately, _unlike_ the `master` branch, the window doesn't attempt to constrain its _aspect ratio_ when resizing. Instead, it letterboxes/pillarboxes the video into the window's area as appropriate using the aforementioned black background. This is a marked improvement, as the aspect-ratio locking was always unnecessary and glitchy.
3. If I resize the window _larger_ than the native resolution, the Source Resolution button appears.
    1. (It _stays_ appeared if I manually resize the preview back down to the native resolution, which is a bit weird but minor.)
4. If I double-click in the window or hit the fullscreen button, the window expands to fill the screen, and the video is shown letterboxed into the 1200px height. So far so good. A second double-click/button press restores the previous size and dimensions.
5. Hitting "Source Resolution" now will, often as not, cause the preview to wrong-size itself slightly within the window after a short delay. (Roughly half a second — moving the mouse outside of the window and causing it to lose focus seems to trigger the wrongsize better, for some reason.) There are now pillarbox bars on either side of the no-longer-1280×720 preview frame.
6. Once the previous has happened, all of the following conditions develop:
    1. I can now freely resize the preview window so that the preview is smaller than the native 1280×720 pixels. The preview image shrinks down to fit the smaller window.
    2. The window **DOES** now constrain its aspect ratio, unlike before. Worse, it's constrained to the _**wrong-sized**_ area where it's always wider than the pillarboxed preview.
    3. If I hit Source Resolution while the preview window is smaller than 1280×720, the _preview_ returns to that size, but the _**window**_ size remains unchanged. I can now can see only the top left corner of the preview. Manually resizing the window immediately causes the preview to resize-to-fit (even if the window is smaller than 1280×720) the moment I start dragging the window border.
    4. Fullscreening a preview window in this state will, the first time, cause the preview to expand into a 1920×1080 image at screen coordinates `(0,0)`, rather than centered on the screen at... um... (finger-math...) `(0,60)` with black bars above and below. Instead, a 120px-tall strip of the desktop is still visible below the preview.
    5. If I leave and then return to fullscreen, it will be correctly centered and will cover the entire display (with letterboxing); the broken fullscreen sizing only happens on the **first** fullscreen attempt after using Source Resolution.
    6. If the preview window was _smaller_ than the native video size when it was fullscreened, then leaving fullscreen the first time will cause it to shrink back to a properly-sized, properly-aspected 1280×720 preview window... that's most likely sticking partway off the bottom or right edge of the screen, because it returns to the original window coordinates but at a much larger size.
    
So, yeah, all a bit screwy. And that's without even involving any larger-than-the-screen videos. (I don't have any of those.)

robxnano commented:
@ferdnyc I think I've fixed most of the problems now. You should be able to resize the window freely, down to a minimum of 640x360, regardless of the video size. I also hid the Source Resolution button in full screen, as well as most cases where the video is already the source resolution. You can still use it when maxmized, which unmaximizes the window in order to do so.

I definitely agree with you re: splitting up the UI file. I'd also like to split some of the overly-large source files like callbacks.c, which seems to have become bloated with lots of unrelated functions which didn't fit elsewhere.

ferdnyc commented:
@robxnano 

> I definitely agree with you re: splitting up the UI file. I'd also like to split some of the overly-large source files like callbacks.c, which seems to have become bloated with lots of unrelated functions which didn't fit elsewhere.

Yeah, `callbacks.c` did feel _really_ "catchall", when I was looking through it.

As far as the UI files go, I'm willing to do the work on splitting those out (actually I've sort of already started), if you like — I can open a PR against your PR branch (still my favorite GitHub trick) with the results when it's done, if that works for you?

If you'd rather keep control of the contents for now while you're still touching parts of `ghb3.ui` (since it'll probably be a pain to keep our branches in sync if we're both messing with it), then I totally get that.

robxnano commented:
@ferdnyc I think it might be better to make those changes on a separate branch rather than here, as the HandBrake team are planning a new release soon and I believe they want to merge this PR beforehand. So I was planning to keep this branch limited to bug fixes in order to not add any complications. I do have another branch [here](https://github.com/robxnano/HandBrake/tree/linux-new-features) with some new features that aren't quite finished yet though, so perhaps you could submit a PR against that one instead.

galad87 commented:
Merged, thanks.

A totally unrelated issue, if you have a chance to take a look: since I added svt-av1 to https://github.com/HandBrake/HandBrake/blob/be9767c7890cc6b61ca5c0bbdb56bc15c27343d8/gtk/src/makedeps.py#L55 and to the following line, the advanced options panel in the video tab is not shown anymore for VP8, VP9, and maybe more, there could be a count somewhere to increase but I didn't find it.

robxnano commented:
@galad87 You just need to change this line: https://github.com/HandBrake/HandBrake/blob/3a3305438027616dcd51092278cac0db16925eed/gtk/src/callbacks.c#L582
You could either increase the limit, or set it to -1 for no limit.

galad87 commented:
Oh. I'll commit a fix soon, thanks again.
